packages/shared/lib/utils/encryption.manager.ts

```diff
@@ -201,31 +201,57 @@ class EncryptionManager extends Encryption {
+            } else {
+                return 'not_started';
+            }
+        } else if (!this.key) {
+            return 'require_decryption';
+        }
 
-        const isEncryptionKeyNew = dbConfig == null && this.key;
-        const isEncryptionIncomplete = dbConfig != null && previousEncryptionKeyHash === encryptionKeyHash && !dbConfig.encryption_complete;
+        const previousEncryptionKeyHash = dbConfig.encryption_key_hash;
+        const encryptionKeyHash = await this.hashEncryptionKey(this.key, this.keySalt);
+        if (previousEncryptionKeyHash !== encryptionKeyHash) {
+            return 'require_rotation';
+        }
+        return dbConfig.encryption_complete ? 'done' : 'incomplete';
+    }
 
-        if (isEncryptionKeyNew || isEncryptionIncomplete) {
-            if (isEncryptionKeyNew) {
-                logger.info('🔐 Encryption key has been set. Encrypting database...');
-                await this.saveDbConfig({ encryption_key_hash: encryptionKeyHash, encryption_complete: false });
-            } else if (isEncryptionIncomplete) {
-                logger.info('🔐 Previously started database encryption is incomplete. Continuing encryption of database...');
-            }
+    public async encryptDatabaseIfNeeded() {
+        const dbConfig = await db.knex.select<DBConfig>('*').from<DBConfig>('_nango_db_config').first();
+        const status = await this.encryptionStatus(dbConfig);
+        const encryptionKeyHash = this.key ? await this.hashEncryptionKey(this.key, this.keySalt) : null;
 
-            await this.encryptDatabase();
-            await this.saveDbConfig({ encryption_key_hash: encryptionKeyHash, encryption_complete: true });
+        if (status === 'disabled') {
             return;
         }
-
-        const isEncryptionKeyChanged = dbConfig?.encryption_key_hash != null && previousEncryptionKeyHash !== encryptionKeyHash;
-        if (isEncryptionKeyChanged) {
-            throw new Error('You cannot edit or remove the encryption key once it has been set.');
+        if (status === 'done') {
+            return;
+        }
+        if (status === 'require_rotation') {
+            throw new Error('Rotation of NANGO_ENCRYPTION_KEY is not supported.');
         }
+        if (status === 'require_decryption') {
+            throw new Error('A previously set NANGO_ENCRYPTION_KEY has been removed from your environment variables.');
+        }
+        if (status === 'not_started') {
+            logger.info('🔐 Encryption key has been set. Encrypting database...');
+            await this.saveDbConfig({ encryption_key_hash: encryptionKeyHash, encryption_complete: false });
+        }
+        if (status === 'incomplete') {
+            logger.info('🔐 Previously started database encryption is incomplete. Continuing encryption of database...');
+        }
+
+        await this.encryptDatabase();
+        await this.saveDbConfig({ encryption_key_hash: encryptionKeyHash, encryption_complete: true });
```

Comment:
At first glance yes, I'm just wondering about the size of the transaction if you have a large db. I think the script assume it can fail at any time and try everything from the start every time which is nice imo